### ISSUE TITLE
Better failed upload procedure

### ISSUE THREAD

**Issue by sanguine-my-brother** [eternia16/javaGMR#18]
When uploading a file to GMR it sometimes fails. However, when one than checks the GMR website it is actually uploaded. This causes a problem that the client keeps showing the game as if it is your turn for a while. 

So I would like to suggest to add a new failed upload procedure which first notifies the user that the upload failed, than giving the user the option to check the GMR site and after that ask the user whether the save actually did upload or not. If it did upload than the game should be removed from the turn games list.

**Comment by laerreal** [comment]
The reason is not a failed uploading. This problem is [known](http://steamcommunity.com/groups/multiplayerrobot/discussions/0/364039785156667718/). Maybe, API server is not updated frequently enough. Consider a parsing of the game web page.

**Comment by sanguine-my-brother** [comment]
I've read that the official client might use an undocumented API call. Therefore, I'm going to run the official client and see what kind of calls this client makes with Wireshark.
Also, I'm not in favour of scraping, but if there're no alternatives then I would suggest web scraping yes.

**Comment by sanguine-my-brother** [comment]
There is indeed an hidden POST endpoint in the official client, multiple actually but related to this context:
http://multiplayerrobot.com/Game/UploadSaveClient

When sending an empty post request I receive the following error message:
`The parameters dictionary contains a null entry for parameter 'turnId' of non-nullable type 'System.Int32' for method 'System.Web.Mvc.ActionResult UploadSaveClient(System.Web.HttpPostedFileBase, Int32, Boolean, System.String)' in 'GmrServer.Controllers.GameController'. An optional parameter must be a reference type, a nullable type, or be declared as an optional parameter.
Parameter name: parameters`

**Comment by laerreal** [comment]
According to captured packed analysis, data of `UploadSaveClient` have to be encapsulated in MIME (like an e-mail).
The structure is:
```
multipart/form-data
    text/plain (turnId)
    text/plain (isCompressed): False - for me
    text/plain (authKey)
    form-data; name=saveFileUpload; filename=[turhId].Civ5Save
```

After, `UploadSaveClient` official client also calls undocumented `GetGamesForPlayer` (not `GetGamesAndPlayers`). The server returns an empty array in JSON. Now content of client
window is empty (probably because I am participating in only one game now).

After a while (about 1 minute), official client ask `GetGamesAndPlayers` and got an updated data.

**Comment by laerreal** [comment]
Also, after `UploadSaveClient` the server sends HTTP OK with `ResultTye` and `PointsEarned` as JSON dictionary.

**Comment by laerreal** [comment]
I think, using of `GetGamesForPlayer` should be sufficient. Send one before uploading of save file, then send it again after uploading. If the game missed then the turn was submitted. Wait for a while and request `GetGamesAndPlayers` again.

**Comment by sanguine-my-brother** [comment]
I created a branch for undocumented upload yesterday, I think I'm going to integrate the other endpoints as well.

**Comment by laerreal** [comment]
A have been made an observation. The server seems to keep responding with outdated information until the game is opened in a browser... Recently, the javaGMR showed old information too long after I sent the save file. But, just after I opened the game web page, javaGMR displayed updated data.

Is it an accidental coincidence? What if the `GetGamesForPlayer` request triggers that logic?

**Comment by sanguine-my-brother** [comment]
Interesting, to say the least, I'll be checking up on this. If this is true then the problem is quite easy to solve. It's trivial to just open the page in the background without the user ever noticing.

**Comment by laerreal** [comment]
I have tried it. Suddenly, `GetGamesForPlayer` seems to have no this effect. Instead, it returns same game list as one returned by `GetGamesAndPlayers`. There is one difference. `GetGamesAndPlayers` wraps game list in dictionary. You can look at `devel` branch in my fork (laerreal/javaGMR).

**Comment by laerreal** [comment]
Hence, it is very likely that `UploadSaveClient` is the right way to upload a turn.

**Comment by sanguine-my-brother** [comment]
So, I've succesfully uploaded a save game through the `UploadSaveClient` endpoint. The uploaded game did no longer appear through the `GetGamesForPlayer` endpoint. 

I would appreciate it if you can confirm that this is working. You can find this code in the `Features/UploadToUndocumentedEndpoint` branch.

**Comment by sanguine-my-brother** [comment]
If this is working that means we can get rid of the piece of code that keeps track of the uploaded games.

**Comment by sanguine-my-brother** [comment]
Comparing version 1.1.1 to the current debugging version. When I upload a game through the `UploadSaveClient` endpoint. The game instantly disappears on the Version 1.1.1 client as well as the official client. So I do think that `UploadSaveClient` is the actual way to go.

**Comment by laerreal** [comment]
Yes. This works like the original client.

**Comment by sanguine-my-brother** [comment]
I've removed the list in jgmrConfig class which keeps track of which games have been uploaded since it is no longer needed. Also, when I had a faulty upload, the game correctly removed it afterwards. Might need to implement a barrier of one minute to prevent double uploads. But other than that it seems to be working perfectly.

**Comment by sanguine-my-brother** [comment]
Marked as closed, going to release version v1.1.2
